### ISSUE TITLE
K64F : TLS handshake fail.

### ISSUE THREAD

**Issue by ashok-rao** [ARMmbed/oob-mbed-os-example-tls#17]
Steps to reproduce:
1. mbed import https://github.com/ARMmbed/oob-mbed-os-example-tls
2. cd oob-mbed-os-example-tls
3. cd tls-client
4. cd mbed-os
5. mbed update mbed-os-5.2.0-rc3
6. cd..
7. mbed compile -t ARM -m K64F
8. copy BUILD\K64F\ARM\tls-client.bin E:\

mbed CLI version = 0.9.9
[EDIT] : K64F connected to network through ethernet.
Open PuTTY with appropriate settings, the TLS handshake fails with the following error. Note different error codes in 3 different trials (resets of target).

![k64_tls_error](https://cloud.githubusercontent.com/assets/17567812/19432514/c1a6ee5a-9454-11e6-8169-b77ec9aebabe.jpg)

**Comment by MarceloSalazar** [comment]
@sbutcher-arm

**Comment by andresag01** [comment]
@sbutcher-arm: I had a brief look into this issue, note that this only occurs with target `ARM` and not `GCC_ARM`. There seem to be two separate problems:

* Sometimes (rather often though) the scheduler throws the error:
```
RTX error code: 0x00000001 task ID: .....
```
I checked the sources for the scheduler and the problem seems to be that we are running out of stack space, which is very unusual because the program seems to work with `gcc`.
```c
#define OS_ERR_STK_OVF          1U
```
I quickly tried to increase the stack size by setting the `OS_MAINSTKSIZE` to `2048` through the `mbed_app.json`, but it does not look like this made a difference.
* The tls-client can never contact the server and always return the `-1` error. It seems that the call that is failing is the user callback `ssl_send()` that is called from mbed TLS when trying to handshake with the server in `mbedtls_ssl_handshake()`. I traced the code manually and it seems that the problem is that the call `_stack->socket_send()` in `TCPSocket::send()` (`mbed-os/features/netsocket/TCPSocket.cpp`) is failing with error value `-3012` (`NSAPI_ERROR_DEVICE_ERROR`). The documentation says:
```c
    NSAPI_ERROR_DEVICE_ERROR  = -3012,     /*!< failure interfacing with the network processor */
```

**Comment by c1728p9** [comment]
The setting ```OS_MAINSTKSIZE``` is misleading because it doesn't actually effect main stack size. The main stack is the buffer ```thread_stack_main``` and is defined in RTX_CM_lib.h. If you double the size there then the stack overflows will go away. This is only useful for debugging though, and won't fix the example since it requires an mbed-os change.

After some digging, it looks like the **stack overflow is indirectly the cause of the TLS handshake fail**. It occurs while a socket connection to the host is open so the socket never gets a chance to gracefully close. On the next boot the same source port is used for the connection which the host rejects since it still thinks the previous session is open. This is why on the first run you see a stack overflow, and on all subsequent runs there is a TLS handshake fail. The error is actually occurring in the connect -```_tcpsocket->connect``` but the return code isn't getting checked. If either the source port is changed (can be done in https://github.com/ARMmbed/mbed-os/blob/master/features/FEATURE_LWIP/lwip-interface/lwip/src/core/lwip_tcp.c#L785) or you wait a couple of minutes then the code will proceed past the handshake and overflow its stack.

As for the stack overflow, it looks like the code is just barely overflowing the 4K main stack. Freeing up 256 bytes should be more than enough to fix the stack overflows.

(Nearly) Wost case stack usage with ARMCC:
```
Function name                                                 Stack usage
-------------------------------------                         -----------
_mutex_acquire() at RTX_CM_lib.h:334                           0      
_malloc_internal()                                             24    
calloc()                                                       16    
mbedtls_mpi_grow() at bignum.c:123                             16    
mbedtls_mpi_copy() at bignum.c:201                             24    
mbedtls_mpi_mul_mpi() at bignum.c:1,176                        56    
mbedtls_mpi_mul_int() at bignum.c:1,215                        0     
mbedtls_mpi_div_mpi() at bignum.c:1,383                        216   
mbedtls_mpi_mod_mpi() at bignum.c:1,455                        0     
mbedtls_mpi_exp_mod() at bignum.c:1,657                        1672  
mbedtls_rsa_public() at rsa.c:302                              40    
mbedtls_rsa_rsassa_pkcs1_v15_verify() at rsa.c:1,355           1088  
mbedtls_rsa_pkcs1_verify() at rsa.c:1,460                      48    
rsa_verify_wrap() at pk_wrap.c:80                              32    
mbedtls_pk_verify() at pk.c:188                                0     
mbedtls_pk_verify_ext() at pk.c:242                            64    
x509_crt_verify_top() at x509_crt.c:1,971                      144   
x509_crt_verify_child() at x509_crt.c:2,105                    224   
mbedtls_x509_crt_verify_with_profile() at x509_crt.c:2,284     0     
mbedtls_ssl_parse_certificate() at ssl_tls.c:4,444             56    
mbedtls_ssl_handshake_client_step() at ssl_cli.c:3,328         120   
mbedtls_ssl_handshake_step() at ssl_tls.c:6,315                0     
mbedtls_ssl_handshake() at ssl_tls.c:6,339                     24    
startTest() at main.cpp:216                                    0     
main() at main.cpp:445                                         216   

Total                                                          4080


```

**Comment by c1728p9** [comment]
I noticed that the biggest consumers of stack - ```mbedtls_rsa_rsassa_pkcs1_v15_verify``` and ```mbedtls_mpi_exp_mod``` have buffers that are configurable. I created #18 and was able to fix the stack overflow and reduce stack usage quite a bit.

**Comment by andresag01** [comment]
@c1728p9: Thanks for having a look at this issue. @yanesca has reviewed the PR, changing the `MBEDTLS_MPI_WINDOW_SIZE` is acceptable and reduces the stack consumption significantly. I have created a PR https://github.com/ARMmbed/oob-mbed-os-example-tls/pull/20 that introduces this change to the example (it has already been merged).

I have tested this fix in the ST Nucleo F429ZI and it seems to fixes the stack overflow problem. However, enabling the mbedtls debug prints still causes a problem. I also need to test this in the K64F.
